fix: use offset-naive datetime in _CredentialsToken

m1racoli opened the pull request:
The field `access_token_acquired_at` is specific to gcloud-aio's Token class and will be used on subsequent calls of `get` with an offset-naive datetime insternally. Therefore we cannot use offset-aware instances of datetime here.

related: #37138 

@Taragolis 

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Commit:
fix: use offset-naive datetime in _CredentialsToken

The field `access_token_acquired_at` is specific to gcloud-aio's Token
class and will be used on subsequent calls of `get` with an offset-naive
datetime insternally. Therefore we cannot use offset-aware instances of
datetime here.

m1racoli committed:
Merge branch 'main' into fix-acquire-access-token

Taragolis commented on airflow/providers/google/common/hooks/base_google.py:
```diff
@@ -674,7 +674,10 @@ async def acquire_access_token(self, timeout: int = 10) -> None:
+        # access_token_acquired_at is specific to gcloud-aio's Token. On subsequent calls of `get` it will be used
+        # with `datetime.datetime.utcnow()`. Therefore we have to use an offset-naive datetime.
+        # https://github.com/talkiq/gcloud-aio/blob/f1132b005ba35d8059229a9ca88b90f31f77456d/auth/gcloud/aio/auth/token.py#L204
+        self.access_token_acquired_at = datetime.datetime.utcnow()  # noqa: TID251
```
noqa not an option here, since this related to the deprecated things in Python 

```suggestion
        self.access_token_acquired_at = datetime.datetime.now(tz=datetime.timezone.utc).replace(tzinfo=None)
```

m1racoli replied:
~~yeah, but without upgrading gcloud-aio-auth's major version (which is a big project on its own) we are forced to use the naive version for now, if we do not want to break existing functionality.~~

Taragolis commented on airflow/providers/google/common/hooks/base_google.py:
```diff
@@ -674,7 +674,10 @@ async def acquire_access_token(self, timeout: int = 10) -> None:
         self.access_token = cast(str, self.credentials.token)
         self.access_token_duration = 3600
-        self.access_token_acquired_at = datetime.datetime.now(tz=datetime.timezone.utc)
+        # access_token_acquired_at is specific to gcloud-aio's Token. On subsequent calls of `get` it will be used
+        # with `datetime.datetime.utcnow()`. Therefore we have to use an offset-naive datetime.
+        # https://github.com/talkiq/gcloud-aio/blob/f1132b005ba35d8059229a9ca88b90f31f77456d/auth/gcloud/aio/auth/token.py#L204
```
Just want to mention, that in 5.2 the type is changed to the aware:
https://github.com/talkiq/gcloud-aio/blob/4979115547cf07bc93878158e86263833e3839fc/auth/gcloud/aio/auth/token.py#L220

It is worthwhile to add this info into the `provider.yaml` for avoid surprises when someone want to relax requirements

https://github.com/apache/airflow/blob/5a0be392e66f8e5426ba3478621115e92fcf245b/airflow/providers/google/provider.yaml#L92

m1racoli replied:
Nice! In terms of avoiding surprises, I've adjusted the test of `get` such that it certainly will make sure that `access_token_acquired_at` is compatible.

m1racoli replied:
When upgrading to gcloud-aio-auth 5 we will override the [refresh](https://github.com/talkiq/gcloud-aio/blob/4979115547cf07bc93878158e86263833e3839fc/auth/gcloud/aio/auth/token.py#L230) method and thus don't have to deal with `access_token_acquired_at` and other boilerplate code anymore.

m1racoli committed:
Update airflow/providers/google/common/hooks/base_google.py

Co-authored-by: Andrey Anshin <Andrey.Anshin@taragol.is>

Commit:
docs: add hints for upgrading gcloud-aio-auth

Let others know about things to consider when upgrading the major
version of gcloud-aio-auth.

eladkal commented on airflow/providers/google/provider.yaml:
```diff
@@ -89,6 +89,12 @@ dependencies:
   - apache-airflow>=2.6.0
   - apache-airflow-providers-common-sql>=1.7.2
   - asgiref>=3.5.2
+  # When upgrading the major version of gcloud-aio-auth we want to make sure to
```
Is something block us from upgrading version now?

m1racoli replied:
Generally speaking, no. I suppose a bit more care is needed when doing so, as it affects a range of async code of the provider. Also not sure yet how it affects the other gcloud-aio packages.
